driver: gpio: dw: Fixing Pin Direction Setting.

ravikh19 opened the pull request:
Change Summary:
Moving the setting of Pin direction before setting/clearing the pin configured as output for the change to correctly take place.

Signed-off-by: Ravik Hasija <ravikh@fb.com>

ravikh19 committed:
driver: gpio: Fixing Pin Direction Setting.

Change Summary:
Moving the setting of Pin direction before setting/clearing the pin
configured as output for the change to correctly take place.

Signed-off-by: Ravik Hasija <ravikh@fb.com>

cfriedt commented on drivers/gpio/gpio_dw.c:
```diff
@@ -243,6 +243,9 @@ static inline void dw_pin_config(const struct device *port,
 	/* Set init value then direction */
 	pin_is_output = (flags & GPIO_OUTPUT) != 0U;
+
+	dw_set_bit(port_base_addr, dir_port, pin, pin_is_output);
```
Is this the port direction register?

ravikh19 replied:
@cfriedt Yes, it is the port direction register. This change basically sets the direction to output first, and then set/clr the respective pin, instead of other way round.

cfriedt commented:
@ravikh19 - just one clarifying question, but otherwise LGTM

ravikh19 commented:
> @ravikh19 - just one clarifying question, but otherwise LGTM

@cfriedt Thanks for the review. I have also updated API to update direction to use base_addr instead of port_base_addr, since port_addr_base will not work for cases when since IP is split into multiple DTS nodes. Also, looks like writing to direction register was the only odd one out that was using port_base_addr. Please have a look & let me know. Thanks.

cfriedt commented:
> > @ravikh19 - just one clarifying question, but otherwise LGTM
> 
> @cfriedt Thanks for the review. I have also updated API to update direction to use base_addr instead of port_base_addr, since port_addr_base will not work for cases when since IP is split into multiple DTS nodes. Also, looks like writing to direction register was the only odd one out that was using port_base_addr. Please have a look & let me know. Thanks. 

I can't say for certain if port base or base should be used because it's hard to say if I have a board i can test this out on.

Just to minimize scope though, i would suggest  keeping the original change and maybe adding a separate commit for port base vs base.

If you're testing on a board, please run through `tests/drivers/gpio`

ravikh19 commented:
> > > @ravikh19 - just one clarifying question, but otherwise LGTM
> > 
> > 
> > @cfriedt Thanks for the review. I have also updated API to update direction to use base_addr instead of port_base_addr, since port_addr_base will not work for cases when since IP is split into multiple DTS nodes. Also, looks like writing to direction register was the only odd one out that was using port_base_addr. Please have a look & let me know. Thanks.
> 
> I can't say for certain if port base or base should be used because it's hard to say if I have a board i can test this out on.
> 
> Just to minimize scope though, i would suggest keeping the original change and maybe adding a separate commit for port base vs base.
> 
> If you're testing on a board, please run through `tests/drivers/gpio`

Sure, Will create a new PR for base_addr change.

ravikh19 committed:
driver: dw: Use base_addr variable to set dir.

Change Summary:
Update Base addr to set direction to use Aligned IP base address instead
of Port Base Address. Port Base Address + offset to direction register
will set incorrect value for Port B,C & D. For ex: In cases when more than
1 port is configured on the same IP, the DTS node for port B will start
at offset 0xC. Calculating the port using Port Base will yield offset of
PORTB DIR register which is at offset 0x10 from Aligned Base, and as a
result will result in setting DIR register of PORTC instead.

Signed-off-by: Ravik Hasija <ravikh@fb.com>

cfriedt commented:
@jmnlcruz - your input here would be very helpful, or perhaps Dirk Brandewie? @tbursztyka ?

I doubt I have any easy way to test any of the proposed changes

cfriedt commented:
@henrikbrixandersen - PTAL - I don't have a board to test this on, personally, but it seems like a valid fix.

henrikbrixandersen reviewed (approved):
LGTM
